src/uct/tcp/sockcm/sockcm_ep.c

```diff
@@ -14,26 +14,242 @@
         } \
     } while (0)
 
+ucs_status_t uct_sockcm_ep_set_sock_id(uct_sockcm_iface_t *iface, uct_sockcm_ep_t *ep)
+{
+    ucs_status_t status;
+    struct sockaddr *dest_addr = NULL;
+
+    UCS_ASYNC_BLOCK(iface->super.worker->async);
+
+    ep->sock_id_ctx = ucs_malloc(sizeof(*ep->sock_id_ctx), "client sock_id_ctx");
+    if (ep->sock_id_ctx == NULL) {
+	status = UCS_ERR_NO_MEMORY;
+	goto out;
+    }
+
+    dest_addr = (struct sockaddr *) &(ep->remote_addr);
+
+    status = ucs_socket_create(dest_addr->sa_family, SOCK_STREAM,
+			       &ep->sock_id_ctx->sock_id);
+    if (status != UCS_OK) {
+	goto out_free;
+    }
+
+    ep->sock_id_ctx->ep = ep;
+    ucs_list_add_tail(&iface->used_sock_ids_list, &ep->sock_id_ctx->list);
+    ucs_debug("ep %p, new sock_id %d", ep, ep->sock_id_ctx->sock_id);
+    status = UCS_OK;
+    goto out;
+
+out_free:
+    ucs_free(ep->sock_id_ctx);
+out:
+    UCS_ASYNC_UNBLOCK(iface->super.worker->async);
+    return status;
+}
+
+ucs_status_t uct_sockcm_send_client_info(uct_sockcm_iface_t *iface, uct_sockcm_ep_t *ep)
+{
+    uct_sockcm_conn_param_t conn_param;
+    uct_sockcm_priv_data_hdr_t *hdr;
+    ssize_t sent_len = 0;
+    ssize_t recv_len = 0;
+    ssize_t offset = 0;
+    int connect_confirm = -1;
+
+    memset(&conn_param.private_data, 0, UCT_SOCKCM_PRIV_DATA_LEN);
+    hdr = &conn_param.hdr;
+
+    /* pack worker address into private data */
+    hdr->length = ep->pack_cb(ep->pack_cb_arg, "eth0",
+                              (void*)conn_param.private_data);
+    if (hdr->length < 0) {
+        ucs_error("sockcm client (iface=%p) failed to fill "
+                  "private data. status: %s",
+                  iface, ucs_status_string(hdr->length));
+        return UCS_ERR_IO_ERROR;
+    }
+    hdr->status = UCS_OK;
+    conn_param.private_data_len = sizeof(uct_sockcm_conn_param_t);
+
+    recv_len = recv(ep->sock_id_ctx->sock_id, (char *) &connect_confirm,
+                    sizeof(int), 0);
+    ucs_debug("recv len = %d\n", (int) recv_len);
+    ucs_debug("connect confirm = %d\n", connect_confirm);
+
+    sent_len = send(ep->sock_id_ctx->sock_id, (char *) &conn_param + offset,
+                    (conn_param.private_data_len - offset), 0);
```

Comment:
maybe send (hdr->length + sizeof(uct_sockcm_priv_data_hdr_t)) as the len?
conn_param.private_data_len can be too long since the actual data written can be shorter.

Reply:
Is that safe to do with this? https://github.com/openucx/ucx/pull/3570/files#diff-7cf03c5fefa318dcaeaf6bf7312d1911R134

Reply:
how is this related..?